src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,252 @@
+        {
+            List<ProjectSystemReferenceInfo> references;
+
+            try
+            {
+                var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+                references = await GetAllReferencesInConfiguredProjectAsync(activeConfiguredProject);
+            }
+            catch
+            {
+                throw new InvalidProjectFileException();
+            }
+
+            return references.ToImmutableArray();
+        }
+
+        private async Task<ConfiguredProject> GetActiveConfiguredProjectByPathAsync(string projectPath)
+        {
+            var unconfiguredProjectPath = FindUnconfiguredProjectByPath(projectPath);
+
+            var activeConfiguredProject = await unconfiguredProjectPath.GetSuggestedConfiguredProjectAsync();
+            if (activeConfiguredProject is null)
+            {
+                throw new InvalidProjectFileException();
+            }
+
+            return activeConfiguredProject;
+        }
+
+        private UnconfiguredProject FindUnconfiguredProjectByPath(string projectPath)
+        {
+            var unconfigProjects = _configuredProject.Services.ProjectService.LoadedUnconfiguredProjects;
+
+            var unconfiguredProject = unconfigProjects.First(project =>
+                StringComparers.Paths.Equals((string)project.FullPath, projectPath));
+
+            return unconfiguredProject;
+        }
+
+        private async Task<List<ProjectSystemReferenceInfo>> GetAllReferencesInConfiguredProjectAsync(ConfiguredProject selectedConfiguredProject)
+        {
+            GetProjectSnapshotAsync(selectedConfiguredProject);
+
+            var references = GetReferences();
+
+            return references;
+        }
+
+        private void GetProjectSnapshotAsync(ConfiguredProject selectedConfiguredProject)
+        {
+            _projectReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _packageReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _assemblyReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+            _sdkReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
+        }
+
+        private List<ProjectSystemReferenceInfo> GetReferences()
+        {
+            List<ProjectSystemReferenceInfo> references = new List<ProjectSystemReferenceInfo>();
+
+            references.AddRange(_projectReferenceHandler.GetReferences());
+            references.AddRange(_packageReferenceHandler.GetReferences());
+            references.AddRange(_assemblyReferenceHandler.GetReferences());
+            references.AddRange(_sdkReferenceHandler.GetReferences());
+
+            return references;
+        }
+
+        public async Task<bool> TryUpdateReferenceAsync(string projectPath, string targetFrameworkMoniker, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
```

Comment:
It passes a TFM, but you ignore it.

Reply:
it seems that the full path to the project is enough. In which cases should we use fullPath + TFM ?

Reply:
this will be removed from the interface  https://github.com/dotnet/roslyn/pull/48990